src/System.Diagnostics.Process/src/System/Diagnostics/ProcessManager.Win32.cs

```diff
@@ -59,4 +59,182 @@ private bool IsMainWindow(IntPtr handle)
             return true;
         }
     }
+
+    internal static partial class NtProcessManager
+    {
+        private static ProcessModuleCollection GetModules(int processId, bool firstModuleOnly)
+        {
+            // preserving Everett behavior.
+            if (processId == SystemProcessID || processId == IdleProcessID)
+            {
+                // system process and idle process doesn't have any modules
+                throw new Win32Exception(Interop.Errors.EFail, SR.EnumProcessModuleFailed);
+            }
+
+            SafeProcessHandle processHandle = SafeProcessHandle.InvalidHandle;
+            try
+            {
+                processHandle = ProcessManager.OpenProcess(processId, Interop.Advapi32.ProcessOptions.PROCESS_QUERY_INFORMATION | Interop.Advapi32.ProcessOptions.PROCESS_VM_READ, true);
+
+                IntPtr[] moduleHandles = new IntPtr[64];
+                GCHandle moduleHandlesArrayHandle = new GCHandle();
+                int moduleCount = 0;
+                for (;;)
+                {
+                    bool enumResult = false;
+                    try
+                    {
+                        moduleHandlesArrayHandle = GCHandle.Alloc(moduleHandles, GCHandleType.Pinned);
+                        enumResult = Interop.Kernel32.EnumProcessModules(processHandle, moduleHandlesArrayHandle.AddrOfPinnedObject(), moduleHandles.Length * IntPtr.Size, ref moduleCount);
+
+                        // The API we need to use to enumerate process modules differs on two factors:
+                        //   1) If our process is running in WOW64.
+                        //   2) The bitness of the process we wish to introspect.
+                        //
+                        // If we are not running in WOW64 or we ARE in WOW64 but want to inspect a 32 bit process
+                        // we can call psapi!EnumProcessModules.
+                        //
+                        // If we are running in WOW64 and we want to inspect the modules of a 64 bit process then
+                        // psapi!EnumProcessModules will return false with ERROR_PARTIAL_COPY (299).  In this case we can't
+                        // do the enumeration at all.  So we'll detect this case and bail out.
+                        //
+                        // Also, EnumProcessModules is not a reliable method to get the modules for a process. 
+                        // If OS loader is touching module information, this method might fail and copy part of the data.
+                        // This is no easy solution to this problem. The only reliable way to fix this is to
+                        // suspend all the threads in target process. Of course we don't want to do this in Process class.
+                        // So we just to try avoid the race by calling the same method 50 (an arbitrary number) times.
```

Comment:
This seems unlikely to be happening in Win7+. Do you have evidence of this on an OS we support?